src/sparse_blas/backends/cusparse/cusparse_scope_handle.hpp

```diff
@@ -0,0 +1,93 @@
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+#ifndef _ONEMKL_SPARSE_BLAS_BACKENDS_CUSPARSE_SCOPE_HANDLE_HPP_
+#define _ONEMKL_SPARSE_BLAS_BACKENDS_CUSPARSE_SCOPE_HANDLE_HPP_
+
+/**
+ * @file Similar to cublas_scope_handle.hpp
+*/
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+// After Plugin Interface removal in DPC++ ur.hpp is the new include
+#if __has_include(<sycl/detail/ur.hpp>) && !defined(ONEAPI_ONEMKL_PI_INTERFACE_REMOVED)
+#define ONEAPI_ONEMKL_PI_INTERFACE_REMOVED
+#endif
+
+#include <thread>
+
+#include "cusparse_error.hpp"
+#include "cusparse_global_handle.hpp"
+#include "cusparse_helper.hpp"
+
+namespace oneapi::mkl::sparse::cusparse {
+
+class CusparseScopedContextHandler {
+    CUcontext original_;
+    sycl::context *placedContext_;
+    sycl::interop_handle &ih;
+    bool needToRecover_;
+
+#ifdef ONEAPI_ONEMKL_PI_INTERFACE_REMOVED
+    static thread_local cusparse_global_handle<ur_context_handle_t> handle_helper;
+#else
+    static thread_local cusparse_global_handle<pi_context> handle_helper;
+#endif
+
+    CUstream get_stream(const sycl::queue &queue);
+    sycl::context get_context(const sycl::queue &queue);
+
+public:
+    CusparseScopedContextHandler(sycl::queue queue, sycl::interop_handle &ih);
+
+    ~CusparseScopedContextHandler() noexcept(false);
+
+    /**
+     * @brief get_handle: creates the handle by implicitly impose the advice
+     * given by nvidia for creating a cusparse_global_handle. (e.g. one cuStream per device
+     * per thread).
+     * @param queue sycl queue.
+     * @return a pair of: cusparseHandle_t a handle to construct cusparse routines; and a CUDA stream
+     */
+    std::pair<cusparseHandle_t, CUstream> get_handle_and_stream(const sycl::queue &queue);
+
+    /// See get_handle_and_stream
+    cusparseHandle_t get_handle(const sycl::queue &queue);
+
+    // This is a work-around function for reinterpret_casting the memory. This
+    // will be fixed when SYCL-2020 has been implemented for Pi backend.
+    template <typename AccT>
+    inline void *get_mem(AccT acc) {
+        auto cudaPtr = ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(acc);
+        return reinterpret_cast<void *>(cudaPtr);
+    }
+
+    template <typename T>
+    inline void *get_mem(T *ptr) {
+        return reinterpret_cast<void *>(ptr);
+    }
```

Comment:
Why not the following? Do things not work as intended in that case?
```suggestion
    template <typename T>
    inline T *get_mem(T *ptr) {
        return ptr;
    }
```

Reply:
You're right, it seemed needed at the time for some reason. I removed the `get_mem` function for USM in https://github.com/oneapi-src/oneMKL/pull/527/commits/81733315974546d853fdd91aae9de5fcd597d38e. It couldn't be used to factorize common functors with the `get_mem` for accessors as I had hoped.